Parameterize over what it means to be an Asset (Coin v.s. Value)

TimSheard opened the pull request:
This PR parameterizes over what kind of assets a transaction can produce. In the old style we used a Coin, and as an alternative to that we introduce the type Value (which is a kind of multi-asset). Both these types are instances of the Val class, so the code has been made parametric over objects in the Val class.

polinavino committed:
Replacing coin with value.

JaredCorduan commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Genesis.hs:
```diff
@@ -102,19 +102,20 @@ data ShelleyGenesis c = ShelleyGenesis
     . sgActiveSlotsCoeff
 
-instance Crypto crypto => ToJSON (ShelleyGenesis crypto) where
+-- TODO JSON for v?
```
I think we can remove this TODO

JaredCorduan commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/LedgerState.hs:
```diff
@@ -633,65 +634,59 @@ genesisState genDelegs0 utxo0 =
 -- | It can be helpful for coin selection.
-txsizeBound :: forall crypto. (Crypto crypto) => Tx crypto -> Integer
+txsizeBound :: forall crypto v. (CV crypto v) => Tx crypto v -> Integer
 txsizeBound tx = numInputs * inputSize + numOutputs * outputSize + rest
```
I think we might be able to get rid of this function altogether. which would mean we could also clean up the transaction type (it stores the input and output bytes so that `txsizeBound` can remove them). we can do this in a separate PR.

JaredCorduan commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/LedgerState.hs:
```diff
@@ -633,65 +634,59 @@ genesisState genDelegs0 utxo0 =
   where
     deregistrations = filter isDeRegKey (toList $ _certs tx)
 
+-- TODO add forge
```
I don't think we need to add all these TODOs in the code

JaredCorduan replied:
a github issue would be more helpful

JaredCorduan commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Utxo.hs:
```diff
@@ -159,20 +159,20 @@ instance
         "Deposit pot must not be negative (post)"
         (\_ st' -> _deposited st' >= 0),
-      let utxoBalance us = _deposited us + _fees us + balance (_utxo us)
+      let utxoBalance us = vplus (vinject $ _deposited us + _fees us) (balance (_utxo us))
```
can we use `<>` to make it a bit more like the original?

JaredCorduan commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Utxo.hs:
```diff
@@ -301,11 +304,18 @@ utxoInductive = do
   ppup' <- trans @(PPUP crypto) $ TRC (PPUPEnv slot pp genDelegs, ppup, txup tx)
 
-  let outputs = Set.toList (eval (rng (txouts txb)))
+  -- TODO check voper
```
can we remove this? or do we need to check something?

JaredCorduan commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Utxo.hs:
```diff
@@ -301,11 +304,18 @@ utxoInductive = do
+  -- TODO add forge errors
+  -- let (Value vls) = _forge txb
+  -- let cids = Map.keys vls
+  -- all (adaID /=) cids  ?! (ForgingAda (Value vls))
```
let's remove this and make an issue instead

polinavino replied:
This is more to do with actually adding forging functionality than just doing the parametrization. Can definitely be removed from here.

JaredCorduan commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Tx.hs:
```diff
@@ -109,51 +110,44 @@ import Shelley.Spec.Ledger.TxData
-
-deriving via
-  AllowThunksIn
-    '[ "txWitsBytes"
```
don't we still want the annotations to be thunks?

polinavino replied:
We couldnt get deriving via to work with `AllowThunksIn` because of lack of ability to derive `Generic` for everything in `Val`. Is this essential?

redxaxder replied:
If the field is lazy but the instance doesn't mark it as allowed, this will trip thunk detection.

If the field is made strict, this will trigger serialization every time a witness set is modified.

One solution would be to figure out a way to get a generic instance for the val. Another would be to write a manual instance.

JaredCorduan commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/TxData.hs:
```diff
@@ -403,40 +452,46 @@ pattern TxIn addr index <-
+    CV crypto v =>
+    {-# UNPACK #-} !BSS.ShortByteString ->
+    v ->
+    TxOut crypto v
```
it seems we can no longer unpack the Word64 from the coin if we want this to be parametric, so the UTxO memory size will be a bit bigger (one word per TxOut?). we can maybe make an issue to measure the difference.

JaredCorduan commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/TxData.hs:
```diff
@@ -534,24 +593,29 @@ data TxBody crypto = TxBody'
-instance Crypto c => HashAnnotated (TxBody c) c
+instance CV c v => HashAnnotated (TxBody c v) c
+
+-- TODO no thunks in bodyBytes is ok?
```
that's a good point, I think we want the annotations to be thunks

JaredCorduan commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/TxData.hs:
```diff
@@ -754,37 +818,32 @@ instance
-    case decompactAddr bs of
-      Just (_ :: Addr crypto) -> pure $ TxOutCompact bs coin
-      Nothing -> cborError $ DecoderErrorCustom "TxOut" "invalid address"
+    addr <- fromCBOR
```
I think we still need to check that the address parses with `decompactAddr`

JaredCorduan commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/UTxO.hs:
```diff
@@ -163,64 +163,64 @@ txouts tx =
-balance :: UTxO crypto -> Coin
-balance (UTxO utxo) = fromIntegral $ Map.foldl' addCoins 0 utxo
+balance :: CV crypto v => UTxO crypto v -> v
+balance (UTxO utxo) = foldr addCoins vzero utxo
```
`foldl'`?

JaredCorduan commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/UTxO.hs:
```diff
@@ -163,64 +163,64 @@ txouts tx =
+balance (UTxO utxo) = foldr addCoins vzero utxo
   where
-    addCoins !b (TxOutCompact _ a) = a + b
+    addCoins (TxOut _ a) b = vplus a b
```
I think we still want the real constructor `TxOutCompact`. and maybe `<>` would be more clear

mrBliss replied:
Yeah, please don't undo https://github.com/input-output-hk/cardano-ledger-specs/pull/1771. 

(Maybe you started working on this with an out-of-date checkout, one that didn't include that PR, and this change got lost while merging/rebasing? If so, we should be careful that no other such changes are in this PR.)

JaredCorduan commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/UTxO.hs:
```diff
@@ -291,14 +292,32 @@ scriptsNeeded u tx =
+      [cert | cert@(DCertPool (RegPool pparams)) <- toList (_certs txb), rewardAcntUsesScript (_poolRAcnt pparams)]
+    mirCertsUsingScripts =
+      [cert | cert@(DCertMir mir) <- toList (_certs txb), any credentialUsesScript (Map.keys (mirRewards mir))]
+-}
```
dead code?

JaredCorduan commented on shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Constants.hs:
```diff
@@ -88,51 +88,54 @@ data Constants = Constants
+    maxReserves = 10000000,
+    genTxRetries = 6,
+    genTxBudget = 10000
+  }
```
let's not indent so that we can see exactly what changed

JaredCorduan commented on shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Core.hs:
```diff
@@ -58,7 +58,8 @@ import Cardano.Crypto.DSIGN.Class (DSIGNAlgorithm (..))
 import qualified Cardano.Crypto.Hash as Hash
 import Control.Iterate.SetAlgebra (eval, (∪), (⋪))
-import Control.Monad (replicateM)
+
```
mystery line! :)

JaredCorduan commented on shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/MetaData.hs:
```diff
@@ -32,10 +32,11 @@ genMetaData ::
+                [ (frequencyTxWithMetaData, genMetaData'),
+                  (100 - frequencyTxWithMetaData, pure (SNothing, SNothing))
+                ]
+     pure(a,b)
```
I think this doesn't actually change anything

JaredCorduan commented on shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Trace/DCert.hs:
```diff
@@ -150,7 +150,7 @@ genDCerts ::
     )
 genDCerts
-  ge@( GenEnv
+   ge@( GenEnv
```
we don't need this

JaredCorduan commented on shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Update.hs:
```diff
@@ -298,11 +298,12 @@ genUpdate
+                             pure (Nothing, [])
+                            )
+                           ]
+                 pure(a,b)
```
also not needed

JaredCorduan commented on shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs:
```diff
@@ -57,18 +62,18 @@ import Shelley.Spec.Ledger.LedgerState
     KeyPairs,
     UTxOState (..),
-    minfeeBound,
+    minfee,
```
:tada:

JaredCorduan commented on shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs:
```diff
@@ -82,6 +87,8 @@ import Shelley.Spec.Ledger.UTxO
     makeWitnessesFromScriptKeys,
     makeWitnessesVKey,
   )
+-- import Shelley.Spec.Ledger.Value (Val (..))
```
can remove

JaredCorduan commented on shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs:
```diff
@@ -173,40 +179,219 @@ genTxRetry
       -------------------------------------------------------------------------
-      let spendingBalance = spendingBalanceUtxo + (sum (snd <$> wdrls)) - deposits + refunds
+      let spendingBalance = vplus spendingBalanceUtxo (vinject $ (sum (snd <$> wdrls)) - deposits + refunds)
+          n = if (Map.size . unUTxO) utxo < 100 then 3 else 0
```
we could add this `100` to the constants or reuse another constant...

JaredCorduan commented on shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs:
```diff
@@ -359,18 +545,20 @@ genWithdrawals
+            genWrdls (Map.toList wdrls)
+          )
+        ]
+    pure ( a, b)
```
another one that didn't change?

JaredCorduan commented on shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Serialisation/Generators.hs:
```diff
@@ -248,17 +252,16 @@ sizedMetaDatum 0 =
+        MD.I <$> arbitrary,
+        MD.B <$> arbitrary,
+        MD.S <$> (T.pack <$> arbitrary)
+      ]
```
just indendation?

JaredCorduan commented on shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Serialisation/Generators.hs:
```diff
@@ -625,12 +628,12 @@ maxMultiSigListLens = 5
+        RequireAllOf <$> resize maxMultiSigListLens (listOf (sizedMultiSig (n-1))),
+        RequireAnyOf <$> resize maxMultiSigListLens (listOf (sizedMultiSig (n-1))),
+        RequireMOf <$> arbitrary <*> resize maxMultiSigListLens (listOf (sizedMultiSig (n-1)))
+      ]
```
just indentation

JaredCorduan commented on shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Serialisation/Generators/Genesis.hs:
```diff
@@ -43,9 +44,16 @@ import Shelley.Spec.Ledger.Keys
 import Test.Shelley.Spec.Ledger.Utils (mkHash)
 
-genShelleyGenesis :: Crypto c => Gen (ShelleyGenesis c)
+-- TODO this is temporary until I look through the Hedgehog docs
```
do we need this now?

JaredCorduan reviewed (commented):
Thank you for this massive undertaking!! All in all it looks good to me, but I left several comments, but it's all small stuff.

TimSheard committed:
wip, trackdown missing key

TimSheard committed:
Fixed the breaking benchmarks. Break was caused because genState was not consistent between different
parts of the generated code. Slight refactoring to make the changes widely useable.
